### DISCUSSION TITLE
Klipper / Moonraker support

### DISCUSSION THREAD

**Discussion by NotExpectedYet** [OctoFarm/OctoFarm#207] [Feature Request]
>This is less of a question than it is an answer. It was brought to my attention on the Voron discord that you were interested in >supporting a direct connection to Klipper via Moonraker's remote API and you were looking for information and >documentation.

>The project is located here and you can find the API documentation here. There is also a hastily thrown together test client that >can be used as a reference. If you are interested in looking at the source for Mainsail, you can find it here.

>I do need to create some documentation about Klipper "printer objects", as these are the items that you generally work with to >query or subscribe to printer state. I will be more than happy to go into more detail of that with you and provide you a list of >the most useful objects that you will want to query.

>If you need to contact me you can do so on Discord. My user name Arksine#0730. If you prefer you can create an issue in >Moonraker's issue tracker, or we can continue the discussion here.

**Comment by qwuille** [comment]
Curious on how this is faring :D

**Reply by NotExpectedYet** [reply]
We're laying some ground work for it currently. Still discussing a lot but it's moving forward now.

**Reply by qwuille** [reply]
Great to hear 😁
Loved using OctoFarm when I used Octoprint. But using Klipper for a good while, it was inevitable to swap to fluidd/mainsail for the extended functions.
Great job!

**Reply by davidzwa** [reply]
We're aiming to support both OctoPrint Klipper (lets call it KlipperOP) which already works right now and MoonRaker Klipper. 
Could you mention the features of mainsail and or fluidd which attract you the most?

**Reply by NotExpectedYet** [reply]
Was just going to ask the same as David, also... mainsail/fluidd apparently offer multi-printer support. I've never personally looked at them but would be interested to hear more about that.

**Reply by qwuille** [reply]
Sorry for the late reply,
What i wish for the most is the same functions as "KlipperOP" offers, a big bonus would be the possibility to expose the macros that are "pre choosen" in i.e fluidd.

In fluidd you can choose wich macros that are visible and not, only the visible ones is needed.

I know for certain that fluidd has multiprinter support, have not used mainsail, but i suppose it has the same possibilites.
Downside is that you do not get that awesome "dashboard" that octofarm provides, you need to "swap" between the printers in fluidd gui.
That function works just perfect and some does not even know it can do this right out of the box.

Again, macros is imho only addition i wish to have access to when you are checking the details of a printer in octofarm.

**Reply by zdesignstudio** [reply]
Curious as to if there was anymore movement on this?

**Reply by NotExpectedYet** [reply]
> Curious as to if there was anymore movement on this?

 My current v1.2 work "may" allow it more easily but it's still a ways off to actually get it implemented properly. It's still one of my focus' though but it will probably go into the V2.0 rewrite as it will be easier when I relay out the data patterns. Not sure at moment and don't really want to give timescales as I'm not getting as much time as I'd like myself to code OctoFarm at the minute.

**Reply by goldjunge91** [reply]
> We're aiming to support both OctoPrint Klipper (lets call it KlipperOP) which already works right now and MoonRaker Klipper.
> Could you mention the features of mainsail and or fluidd which attract you the most?


From my point I don’t see in a total view a difference between KlipperOP and Mainsail/fluidd
All three are web interface’s to control in some behavior a single printer. 
features that are nice to have:
1. All General controlling functions for the printer
2. Restart Klipper 
3. Restart host 
4. Auto re/connect 
5. Data logging Up/Down times, print stats, printer stats,data log input shaper,
6. Backup for .cfg ( or select files in the config folder) config files
7. Compare printer.cfg (all other .cfg or select )configs from different printers 
8. Update .cfg config for multiple printer in one ( uploaded files )
9. Version control for printer.cfg 
10. Automatically print start (from queue) depends on print volume of the free available printers and Job size.
11. Control over what printers can print for filament based on definition of material in startup code

this are some ideas that I have I my mind of daily dealing with multiple different printers.

**Comment by NotExpectedYet** [comment]
Please see https://github.com/OctoFarm/OctoFarm/discussions/1298
